homeassistant/components/sensor/filter.py

```diff
@@ -156,7 +161,32 @@ def filter_sensor_state_listener(entity, old_state, new_state):
                 self._unit_of_measurement = new_state.attributes.get(
                     ATTR_UNIT_OF_MEASUREMENT)
 
-            self.async_schedule_update_ha_state()
+            if update_ha:
+                self.async_schedule_update_ha_state()
+
+        if 'recorder' in self.hass.config.components:
+            history_list = []
+            for filt in self._filters:
+                if isinstance(filt.window_size, int):
+                    filter_history = await self.hass.async_add_job(partial(
+                        history.get_last_state_changes, self.hass,
+                        filt.window_size, entity_id=self._entity))
+                elif isinstance(filt.window_size, datetime.timedelta):
+                    start = dt_util.utcnow() - filt.window_size
+                    filter_history = await self.hass.async_add_job(partial(
+                        history.state_changes_during_period, self.hass,
+                        start, entity_id=self._entity))
+                for _, states in filter_history.items():
+                    history_list.extend(
+                        [st for st in states if st not in history_list])
```

Comment:
This is very very inefficient. We should not do a query per filter. We should be able to find the size of what to query _before_ we do a SQL query.

I am also a bit confused that we need to do instance checks on `window_size`. Why does a variable have 2 different meanings?

Reply:
It is in fact very inefficient. Unfortunately you can't use set() with Ha.State object (since it is not hashable), that would be more efficient.

You can't know the size in advance because sensors are not always periodic.

The window size is one and only dimension, issue is the scale by which we measure the dimension (time vs # events). It like having weight and supporting pounds and kilos, the difference being that conversion between time and #events is not linear (because #events aren't always periodic)

Reply:
code wise it will be more complex (error prone) for a marginal efficiency gain at startup.

But I will implement that tonight